src/Features/Core/Portable/Workspace/ProjectCacheService.cs

```diff
@@ -21,7 +23,7 @@ namespace Microsoft.CodeAnalysis.Host
     /// cache is likely to outweigh the benefit (for example, in Misc File Workspace cases, we can end up holding
     /// onto a lot of memory even after a file is closed).  We can opt in other kinds of Workspaces as needed.
     /// </summary>
-    internal partial class ProjectCacheService : IProjectCacheHostService
+    internal partial class ProjectCacheService : IProjectCacheHostService, IDisposable
```

Comment:
i woudl still *much* rather have a way for either a Workspace to notify it is going away, or a service you can get from teh a specific workspae to notify that.  and then this type just hooks to that.  this being idisposable, and needing a new way to handle all that disposal logic through our entire system just seems like a lot of complexity i cannot wrap my head around.  once again, this gets to the position of "now i have N+1 ways of doing things, and i cannot figure out what i shoudl use at any point in time".

Reply:
This would require each service that needs to handle workspace disposal to modify the workspace service factory to hook up an event listener, and still effectively use an `IDisposable` path. Alternatives have been attempted in the past (e.g. `IEventListenerStoppable`), but they proved more complicated than the approach here. The new approach for handling MEF-provided workspace services is modeled directly after the ownership behavior of all other MEF-provided objects. This allows substantial simplification as follow-up changes, but those changes are being left for the main branch instead of this release branch.

Reply:
> This would require each service that needs to handle workspace disposal to modify the workspace service factory to hook up an event listener,

This seems fine.  

> and still effectively use an IDisposable path

I don't know what this means.  It's not an "IDisposable" path to me.  It's a "workspace is going away" path for me.  For example, i would not expect anyone to take these services and put them in a using block.  

> The new approach for handling MEF-provided workspace services is modeled directly after the ownership behavior of all other MEF-provided objects. 

We already have a model for mef provided objects through IThreadingContext.DisposalToken.  It's still not clear ot me why that approach is not *sufficient* to unblock our insertion. 

>  This allows substantial simplification as follow-up changes, but those changes are being left for the main branch instead of this release branch.

I would rather us start simple, and then add more complexity in the future rather than start complex and hope we substantially simplify things at a later point.  this part of the system is already nearly incomprehensible.  I'm actually at the point (both before, and esp after) this PR where i literally have no idea at all which of any of our systems here we should be using for this sort of issue.

Reply:
>  I'm actually at the point (both before, and esp after) this PR where i literally have no idea at all which of any of our systems here we should be using for this sort of issue.

Prior to this change, there was no system for this. Services that needed notification from the workspace were directly coupled to the workspace on a one-off basis (`IEventListenerStoppable` was the one-off coupling mechanism for solution crawler). After this change, there is exactly one system for this, which is the one in this pull request. The subsequent cleanup involves decoupling services by transitioning to this.

> We already have a model for mef provided objects through IThreadingContext.DisposalToken.

This is a totally different context. We could switch to only using this disposal token if and only if we could remove `MefWorkspaceServices` at the same time (i.e. the reason why we need the concepts separate is exactly the same reason why the concepts have always been separated).